### PR TITLE
Qt settings [14957]

### PR DESCRIPTION
**Pull request by jparisu**
- #14 

This PR support Qt settings, so every time the app is opened, it gets the last configuration used and stored in local qt data.

### CODE REVIEW

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    QDomElement stamp_elem = doc.createElement(USE_HEADER_STAMP_SETTINGS_TAG);
+    stamp_elem.setAttribute("value", data_type_configuration.use_header_stamp ? "true" : "false");
+    parent_element.appendChild(stamp_elem);
+
+    QDomElement discard_elem = doc.createElement(DISCARD_LARGE_ARRAYS_SETTINGS_TAG);
+    discard_elem.setAttribute("value", data_type_configuration.discard_large_arrays ? "true" : "false");
+    parent_element.appendChild(discard_elem);
+
+    return true;
```
Comment (antonrv) [inline plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: Why we need to return a boolean if always return true? Any future scenario returning false? Same for `xmlLoadState` method

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    max_elem.setAttribute("value", QString::number(data_type_configuration.max_array_size));
+    parent_element.appendChild(max_elem);
+
+    QDomElement stamp_elem = doc.createElement(USE_HEADER_STAMP_SETTINGS_TAG);
+    stamp_elem.setAttribute("value", data_type_configuration.use_header_stamp ? "true" : "false");
+    parent_element.appendChild(stamp_elem);
+
+    QDomElement discard_elem = doc.createElement(DISCARD_LARGE_ARRAYS_SETTINGS_TAG);
+    discard_elem.setAttribute("value", data_type_configuration.discard_large_arrays ? "true" : "false");
+    parent_element.appendChild(discard_elem);
```
Comment (antonrv) [inline plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: Where is data_type_configuration coming from? I cannot find its definition
Does it make sense for `xmlSaveState` and `xmlLoadState` to be static class functions?

File: plugins/datastreamer_plugin/CMakeLists.txt [hunk 3]
```diff
@@ -110,6 +110,7 @@ set (ALL_LIBRARY_SOURCES
         # UI files
         "ui/topic_selection_dialog/dialogselecttopics.cpp"
         "ui/topic_selection_dialog/dialogselecttopics.h"
+        "ui/topic_selection_dialog/Configuration.cpp"
```
Comment (antonrv) [inline plugins/datastreamer_plugin/CMakeLists.txt]: Only referring to header *.h files: are they really necessary here? Would cmake work just fine without them?

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+{
+    // Get default Qt settings
+    QSettings settings;
+
+    settings.setValue(prefix + MAX_ARRAY_SIZE_SETTINGS_TAG, data_type_configuration.max_array_size);
+    settings.setValue(prefix + USE_HEADER_STAMP_SETTINGS_TAG, data_type_configuration.use_header_stamp);
+    settings.setValue(prefix + DISCARD_LARGE_ARRAYS_SETTINGS_TAG, data_type_configuration.discard_large_arrays);
+    settings.setValue(prefix + XML_DATATYPE_FILES_SETTINGS_TAG, xml_datatypes_files);
+    settings.setValue(prefix + DOMAIN_ID_SETTINGS_TAG, domain_id);
+}
```
Comment (antonrv) [inline plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: A new `QSettings` object is being built and destroyed at the end of the scope, what for?

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        const QString& prefix)
+{
+    // Get default Qt settings
+    QSettings settings;
+
+    data_type_configuration.max_array_size = settings.value(prefix + MAX_ARRAY_SIZE_SETTINGS_TAG, data_type_configuration.max_array_size).toInt();
+    data_type_configuration.use_header_stamp = settings.value(prefix + USE_HEADER_STAMP_SETTINGS_TAG, data_type_configuration.use_header_stamp).toBool();
+    data_type_configuration.discard_large_arrays = settings.value(prefix + DISCARD_LARGE_ARRAYS_SETTINGS_TAG, data_type_configuration.discard_large_arrays).toBool();
+    xml_datatypes_files = settings.value(prefix + XML_DATATYPE_FILES_SETTINGS_TAG, xml_datatypes_files).toStringList();
+    domain_id = settings.value(prefix + DOMAIN_ID_SETTINGS_TAG, domain_id).toInt();
```
Comment (antonrv) [inline plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: Some clarification is needed here to know the effects of this function and where `data_type_configuration` is defined

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.hpp [hunk 6]
```diff
@@ -42,6 +42,14 @@ namespace ui {
 struct Configuration
 {
 
+    Configuration();
```
Comment (antonrv) [inline plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.hpp:41]: NIT: Referring to `struct Configuration` doxygen (even though not belongs to this PR):

`every Fast DDS entity required` -> `every required Fast DDS entity`

Minor grammar typo and not clear what `the process requires to instantiate` means:
`It create, manage and destroy every Fast DDS entity that the process requires to instantiated` -> `It creates, manages and destroys every required Fast DDS entity.`?

**Review by antonrv — CHANGES_REQUESTED** [review]
Just some clarifications needed

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp (hunk 1 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: This is not perse an overriden method, but it is the equivalent to `FastDdsDataStreamer::xmlSaveState`, that it is indeed an overriden method, and it has to return a boolean.

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp (hunk 2 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: Sorry, this has lose coherence in a rebase, this branch must be after #14 .
Now it should make more sense.

File: plugins/datastreamer_plugin/CMakeLists.txt (hunk 3 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/CMakeLists.txt]: I do not think they are necessary. I copy pasted from `plotjuggler-ros-plugins` and it does it this way, should I try to change it in this PR or we open a new one?

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp (hunk 4 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: This is the basic usage of QSettings. Creating a new object of this class stores these values in the Qt persistent data.
I added the name of organization an app to be more precise where to store it.

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp (hunk 5 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.cpp]: again, rebase problem. Sould be fixed already

File: plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.hpp (hunk 6 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/ui/topic_selection_dialog/Configuration.hpp:41]: this was a copy/pasted doxygen, it was incorrect.

**Review by antonrv — APPROVED** [review]

**Pull request merged**